docs/dev/guidelines/server-tests.rst

```diff
@@ -0,0 +1,180 @@
+
+
+2. ArchUnit
+===========
+Use the `ArchUnit <https://www.archunit.org/>`__ library to prevent the unintentional inclusion of unnecessary packages. We use the library to enforce consistency in the code base.
+Here is a simple ArchUnit test using an ArchRule to forbid JUnit assertions (in favor of AssertJ ones).
+
+.. code-block:: java
+
+    @Test
+    void testNoJunitJupiterAssertions() {
+        ArchRule noJunitJupiterAssertions = noClasses().should().dependOnClassesThat().haveNameMatching("org.junit.jupiter.api.Assertions");
+
+        noJunitJupiterAssertions.check(testClasses);
+    }
+
+We first define the ArchRule prohibiting the JUnit assertion package and then enforce it in test classes.
+
+
+3. Counting database query calls within tests
+==============================================
+
+It's possible to write tests checking how many database accesses a REST call performs. These tests ensure that code changes don't lead to more database calls or remind developers if they do, which is especially important for commonly used endpoints.
+However, we should carefully consider before adding such assertions as it makes the test more tedious to maintain.
+
+The test below tracks how many database accesses a REST call performs. The custom assert ``assertThatDb`` uses the ``HibernateQueryInterceptor`` to count the number of queries. The assertion checks the number of database accesses and returns the original result of the REST call, which you can continue to use throughout the test.
+
+.. code-block:: java
+
+    class TestClass {
+
+        @Test
+        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+        void testQueryCount() throws Exception {
+            Course course = assertThatDb(() -> request.get("/api/courses/" + courses.get(0).getId() + "/for-dashboard", HttpStatus.OK, Course.class)).hasBeenCalledTimes(3);
+            assertThat(course).isNotNull();
+        }
+    }
+
+4. UtilServices and factories
+=============================
+When setting up data in tests, use helper functions from corresponding UtilService and Factory classes. We use the factory method pattern to structure test cases, depicted in the simplified class
+diagram below. In this example, the ``ExerciseTestFactory`` generates exercises for tests, while the ``FileUploadTestFactory`` specializes in file upload exercises. The ``FileUploadIntegrationTest``
+utilizes the services provided by ``FileUploadTestService``. The ``FileUploadTestService`` manages the creation and communication with the database.
+
+.. figure:: resources/FileUpload_UtilService_Factory.png
+    :align: center
+    :alt: File upload UtilService and Factory
+
+
+In general, **UtilServices** manage the communication with the database, and **Factories** are responsible for object creation and initialization. If you cannot find the correct helper function, add a new one to the most fitting UtilService or Factory and enhance it with JavaDoc.
```

Comment:
Can you add a small note to this section to always use the ids retuned by the database and not to assume the existence or non-existence of specific values.

Reply:
added a note